packages/types/types/src/record.ts

```diff
@@ -3,45 +3,102 @@
  * Licensed under the MIT License.
  */
 
+/** An observable key/value collection, such as a Record or Map. */
+export interface IShapeProducer<TKey> {
+    /**
+     * Acquire a reader for this producer's values and implicitly subscribe the consumer
+     * to value change notifications.
+     *
+     * @param consumer - The consumer to be notified of value changes.
+     */
+    open(consumer: IShapeConsumer<TKey>): IShapeReader<TKey>;
+
+    /**
+     * Unsubscribe the given 'consumer' from this producer's change notifications.
+     *
+     * @param consumer - The consumer to unregister from the producer.
+     */
+    close(consumer: IShapeConsumer<TKey>): void;
+}
+
+/** A consumer of change notifications for a key/value collection, such as a Record or Map. */
+export interface IShapeConsumer<TKey> {
+    keyChanged(key: TKey, producer: IShapeProducer<TKey>): void;
+}
+
+/** Capability to read values of a key/value collection. */
+export interface IShapeReader<TKey> {
+    keys(): IterableIterator<TKey>;
+
+    has(key: TKey): boolean;
+
+    readonly size: number;
+
+    /**
+     * A reference to the underlying producer that provides the shape for this reader,
+     * or undefined if the producer is immutable.
+     */
+    readonly producer?: IShapeProducer<TKey>;
+}
+
+export interface IShapeWriter<TKey> {
+    delete(key: TKey): void;
+    clear(): void;
+}
+
 /** An observable key/value collection, such as a Record or Map. */
 export interface IProducer<T> {
```

Comment:
Our base 'IProducer' is the only variant that isn't implicitly an IShapeProducer.

IMapProducer (below) is a convenient way to implement both IProducer and IShapeProducer